File: download/src/main/java/com/huison/download/DownloadDBCenter.java
package com.huison.download;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.huison.download.utils.CloseUtil;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by huisonma on 2019/5/15.
 */
class DownloadDBCenter {

    private DownloadDBHelper dbHelper;

    DownloadDBCenter(Context context, String dir) {
        File file = new File(dir);
        if (!file.exists()) {
            file.mkdirs();
        }
        dbHelper = new DownloadDBHelper(context, dir);
    }

    private static ContentValues toContentValues(DownloadComponent component) {
        ContentValues values = new ContentValues();
        values.put("url", component.getUrl());
        values.put("path", component.getPath());
        values.put("startByte", component.getStart());
        values.put("endByte", component.getEnd());
        values.put("number", component.getNumber());
        values.put("downloadedLength", component.getDownloadedLength());
        values.put("totalLength", component.getTotalLength());
        return values;
    }

    boolean saveComponent(DownloadComponent component) {
        if (component == null) {
            return false;
        }
        if (hasDownloaded(component)) {
            return updateComponent(component) != -1;
        } else {
            return insertComponent(component) != -1;
        }
    }

    private boolean hasDownloaded(DownloadComponent component) {
        if (component == null) {
            return false;
        }
        SQLiteDatabase database = dbHelper.getReadableDatabase();
        boolean result = false;
        Cursor cursor = null;
        try {
            String[] whereArgs = {component.getUrl(), String.valueOf(component.getNumber())};
            cursor = database.query(
                    DownloadDBHelper.TABLE_NAME,
                    new String[]{"url", "number"},
                    "url = ? and number = ?",
                    whereArgs,
                    null, null, null);
            result = cursor != null && cursor.getCount() > 0;
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            CloseUtil.close(cursor);
        }
        return result;
    }

    private long insertComponent(DownloadComponent component) {
        long rowId = -1;
        if (component == null) {
            return rowId;
        }
        try {
            rowId = dbHelper.getWritableDatabase().insert(
                    DownloadDBHelper.TABLE_NAME,
                    null,
                    toContentValues(component));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return rowId;
    }

    private int updateComponent(DownloadComponent component) {
        if (component == null) {
            return -1;
        }
        try {
            SQLiteDatabase database = dbHelper.getWritableDatabase();
            String[] whereArgs = {component.getUrl(), String.valueOf(component.getNumber())};
            return database.update(
                    DownloadDBHelper.TABLE_NAME,
                    toContentValues(component),
                    "url = ? and number = ?",
                    whereArgs);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return -1;
    }

    List<DownloadComponent> queryComponents(String url) {
        SQLiteDatabase database = dbHelper.getReadableDatabase();
        Cursor cursor = null;
        List<DownloadComponent> components = new ArrayList<>();
        try {
            cursor = database.query(
                    DownloadDBHelper.TABLE_NAME,
                    null,
                    "url = ?",
                    new String[]{url},
                    null, null, "number");
            while (cursor.moveToNext()) {
                String path = cursor.getString(cursor.getColumnIndex("path"));
                long start = cursor.getLong(cursor.getColumnIndex("startByte"));
                long end = cursor.getLong(cursor.getColumnIndex("endByte"));
                int threadId = cursor.getInt(cursor.getColumnIndex("number"));
                long progress = cursor.getLong(cursor.getColumnIndex("downloadedLength"));
                long contentLength = cursor.getLong(cursor.getColumnIndex("totalLength"));
                DownloadComponent component = new DownloadComponent(url, path, start, end, threadId, progress, contentLength);
                components.add(component);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            CloseUtil.close(cursor);
        }
        return components;
    }

    int deleteComponents(String url) {
        try {
            return dbHelper.getWritableDatabase().delete(
                    DownloadDBHelper.TABLE_NAME,
                    "url=?",
                    new String[]{url});
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 0;
    }

    int deleteAllComponents() {
        try {
            return dbHelper.getWritableDatabase().delete(DownloadDBHelper.TABLE_NAME, null, null);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 0;
    }
}

File: ktdownload/src/main/java/com/huison/kotlin/download/DownloadDBCenter.kt
package com.huison.kotlin.download

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import java.io.File

/**
 * Created by huisonma 2019/9/3.
 */
class DownloadDBCenter internal constructor(context: Context, dir: String) {

    companion object {

        const val TABLE_NAME = DownloadDBHelper.TABLE_NAME

        fun toContentValues(component: DownloadComponent): ContentValues {
            val values = ContentValues()
            values.put("url", component.url)
            values.put("path", component.path)
            values.put("startByte", component.start)
            values.put("endByte", component.end)
            values.put("number", component.number)
            values.put("downloadedLength", component.downloadedLength)
            values.put("totalLength", component.totalLength)
            return values
        }
    }

    private val database: SQLiteDatabase

    init {
        val file = File(dir)
        if (!file.exists()) {
            file.mkdirs()
        }
        database = DownloadDBHelper(context, dir).writableDatabase
    }

    internal fun saveComponent(component: DownloadComponent): Boolean {
        if (hadDownload(component)) return updateComponent(component) != -1

        return insertComponent(component) != -1L
    }

    private fun hadDownload(component: DownloadComponent?): Boolean {
        if (component == null) return false

        val cursor = database.query(TABLE_NAME,
                arrayOf("url", "number"),
                "url = ? and number = ?",
                arrayOf(component.url, component.number.toString()),
                null, null, null)

        val result: Boolean = cursor?.count!! > 0
        cursor.close()
        return result
    }

    private fun insertComponent(component: DownloadComponent?): Long {
        if (component == null) return -1
        return database.insert(TABLE_NAME, null, toContentValues(component))
    }

    private fun updateComponent(component: DownloadComponent?): Int {
        if (component == null) return -1
        return database.update(TABLE_NAME, toContentValues(component), "url = ? and number = ?",
                arrayOf(component.url, component.number.toString()))
    }


    internal fun queryComponents(url: String): List<DownloadComponent> {
        val components = ArrayList<DownloadComponent>()
        val cursor = database.query(TABLE_NAME, null, "url = ?", arrayOf(url), null, null, "number")
        while (cursor.moveToNext()) {
            val path = cursor.getString(cursor.getColumnIndex("path"))
            val start = cursor.getLong(cursor.getColumnIndex("startByte"))
            val end = cursor.getLong(cursor.getColumnIndex("endByte"))
            val number = cursor.getInt(cursor.getColumnIndex("number"))
            val downloadedLength = cursor.getLong(cursor.getColumnIndex("downloadedLength"))
            val totalLength = cursor.getLong(cursor.getColumnIndex("totalLength"))
            val component = DownloadComponent(url, path, start, end, number, downloadedLength, totalLength)
            components.add(component)
        }
        cursor.close()
        return components
    }

    internal fun deleteComponents(url: String?): Int {
        if (url == null) return 0
        return database.delete(TABLE_NAME, "url = ?", arrayOf(url))
    }

    internal fun deleteAllComponents(): Int {
        return database.delete(TABLE_NAME, null, null)
    }
}